index.js

```diff
@@ -125,6 +136,460 @@
+      }
+    }
+    return upcarrots + '\n' + highlightNumbers;
+  };
+
+  //  sigGetName :: Node -> String
+  var sigGetName = function(sig) {
+    return sig.children[SIG_IDX_NAME].children[0].text;
+  };
+
+  //  findConstraint :: (TypeClass, String) -> Node -> Nullable Node
+  var findConstraint = function(typeClass, typeVarName) {
+    return function(sig) {
+      return traverseFind(function(node, parents) {
+        if (isVar(node) && node.text === typeVarName) {
+          var parent = getParent(parents);
+          if (parent.text === getConstraintText(typeClass)) {
+            return parent;
+          }
+        }
+      }, sig.children[SIG_IDX_CONSTRAINTS]);
+    };
+  };
+
+  //  findTypeVar :: (String, Node) -> Nullable Node
+  var findTypeVar = function(typeVarName, startNode) {
+    return traverseFind(function(node, parents) {
+      if (isVar(node) && node.text === typeVarName) {
+        return node;
+      }
+    }, startNode);
+  };
+
+  //  findArg :: (Integer, Node) -> Node
+  var findArg = function(index, sig) {
+    var args = nonDelimiters(sig.children[SIG_IDX_ARGS].children);
+    return args[index];
+  };
+
+  //  findArgTypeVar :: (Integer, String) -> Node -> Nullable Node
+  var findArgTypeVar = function(index, typeVarName) {
+    return function(sig) {
+      return findTypeVar(typeVarName, findArg(index, sig));
+    };
+  };
+
+  //  findVal :: Integer -> Node -> Node
+  var findVal = function(index) {
+    return function(sig) {
+      return isNaN(index) ? sig.children[SIG_IDX_RET] :
+                            findArg(index, sig);
+    };
+  };
+
+  //  findRetTypeVar :: String -> Node -> Node
+  var findRetTypeVar = function(typeVarName) {
+    return function(sig) {
+      return findTypeVar(typeVarName, sig.children[SIG_IDX_RET]);
+    };
+  };
+
+  //  pipe :: (Any -> [(Any -> Any)]) -> Any
+  var pipe = function(data, fns) {
+    var r = fns[0](data);
+    for (var idx = 1; idx < fns.length; idx += 1) {
+      r = fns[idx](r);
+    }
+    return r;
+  };
+
+  //  head :: [a] -> a?
+  var head = function(xs) {
+    return xs[0];
+  };
+
+  //  tail :: [a] -> [a]
+  var tail = function(xs) {
+    return xs.slice(1);
+  };
+
+  //  last :: [a] -> a?
+  var last = function(xs) {
+    return xs[xs.length - 1];
+  };
+
+  //  toPairs :: StrMap a -> [Pair String a]
+  var toPairs = function(obj) {
+    var pairs = [];
+    for (var k in obj) {
+      pairs.push([k, obj[k]]);
+    }
+    return pairs;
+  };
+
+  //  isFirstChild :: (Node, Node) -> Boolean
+  var isFirstChild = function(node, parent) {
+    return head(nonDelimiters(parent.children)) === node;
+  };
+
+  //  isRootFirstChild :: (Node, [Node]) -> Boolean
+  var isRootFirstChild = function(node, parents) {
+    return (
+      isRootChild(parents) &&
+      isFirstChild(node, head(parents))
+    );
+  };
+
+  //  getParent :: [Node] -> Nullable Node
+  var getParent = last;
+
+  //  isRoot :: [Node] -> Boolean
+  var isRoot = isEmpty;
```

Comment:
> the answer may be the offending occurrence and all previous occurrences, since it's the set of types common to all previous a values which constrain subsequent a values.

Okay, hopefully that won't be too hard since that is what I originally set out to do.

> How can we represent the a at position 1 (or any other position) as a single value? Shouldn't we list all the array's values?

Good questions, the current result is somewhat awkward. I feel like the value could help a developer quickly identify exactly what has gone wrong, seems a shame to lose it. It is just a string, maybe we can truncate it? Still have to figure out how we'd highlight it if we're going to show multiple values. Maybe a special case for "collection" (Array, StrMap) types? Gotta run, will check back later.